### PR TITLE
refactor: 重构档案系统

### LINKED ISSUE #17: [Feature] 支持 Yggdrasil Connect
### 检查项

- [X] 我已在 [Issues 页面](https://github.com/Hex-Dragon/PCL2/issues?q=is%3Aissue+) 和 [常见&难检反馈及问题列表](https://github.com/Hex-Dragon/PCL2/discussions/1930) 中搜索，确认了这一提案未被提交过。
- [X] 我已查看 [功能投票页面](https://github.com/Hex-Dragon/PCL2/discussions/categories/%E5%8A%9F%E8%83%BD%E6%8A%95%E7%A5%A8/)，确认了这一提案未在投票列表中。
- [X] 我知晓还没做的新功能真的太多了，忙不过来，所以新功能提案几乎不会被处理，也不建议再提交新功能提案 qwq……

### 描述

```
{
  "meta": {
    "serverName": "LittleSkin",
    "implementationName": "Yggdrasil Connect",
    "implementationVersion": "0.0.6",
    "links": {
      "announcement": "https://littleskin.cn/api/announcements",
      "homepage": "https://littleskin.cn",
      "register": "https://littleskin.cn/auth/register"
    },
    "feature.no_email_login": true,
    "feature.openid_configuration_url": "https://open.littleskin.cn/.well-known/openid-configuration"
  },
  "skinDomains": [
    "littleskin.cn"
  ],
  "signaturePublickey": "-----BEGIN PUBLIC KEY-----\nMIICIjANBgkqhkiG9w0BAQEFAAOCAg8AMIICCgKCAgEArGcNOOFIqLJSqoE3u0hj\ntOEnOcET3wj9Drss1BE6sBqgPo0bMulOULhqjkc/uH/wyosYnzw3xaazJt87jTHh\nJ8BPMxCeQMoyEdRoS3Jnj1G0Kezj4A2b61PJJM1DpvDAcqQBYsrSdpBJ+52MjoGS\nvJoeQO5XUlJVQm21/HmJnqsPhzcA6HgY71RHYE5xnhpWJiPxLKUPtmt6CNYUQQoS\no2v36XWgMmLBZhAbNOPxYX+1ioxKamjhLO29UhwtgY9U6PWEO7/SBfXzyRPTzhPV\n2nHq7KJqd8IIrltslv6i/4FEM81ivS/mm+PN3hYlIYK6z6Ymii1nrQAplsJ67OGq\nYHtWKOvpfTzOollugsRihkAG4OB6hM0Pr45jjC3TIc7eO7kOgIcGUGUQGuuugDEz\nJ1N9FFWnN/H6P9ukFeg5SmGC5+wmUPZZCtNBLr8o8sI5H7QhK7NgwCaGFoYuiAGL\ngz3k/3YwJ40BbwQayQ2gIqenz+XOFIAlajv+/nyfcDvZH9vGNKP9lVcHXUT5YRnS\nZSHo5lwvVrYUrqEAbh/zDz8QMEyiujWvUkPhZs9fh6fimUGxtm8mFIPCtPJVXjeY\nwD3Lvt3aIB1JHdUTJR3eEc4eIaTKMwMPyJRzVn5zKsitaZz3nn/cOA/wZC9oqyEU\nmc9h6ZMRTRUEE4TtaJyg9lMCAwEAAQ==\n-----END PUBLIC KEY-----\n"
}
```

### 原因

LittleSkin 在 Yggdrasil API 添加了用于发现 OAuth 认证服务器的配置，应该能实现更优秀的体验

### LINKED ISSUE #2867: 同步上游的正版令牌获取机制
### 检查项

- [x] 我已在 [Issues 页面](https://github.com/PCL-Community/PCL2-CE/issues?q=is%3Aissue+) 中搜索，确认了这一建议未在社区版被提交过
- [ ] 我确认只是全部选中而没有[仔细确认](https://github.com/PCL-Community/PCL2-CE/issues?q=is%3Aissue+)就直接提交了这个 Issue，并且同意这个 Issue 可以直接被关闭
- [x] 我确认我所描述的改进是现实的、有理有据的，而不是我临时起意的
- [x] 我已知悉社区不对这类 Issue 的完成时间作出任何保证，并有足够的耐心等待社区开发者的处理，同时已知悉如果我作出催促、质问或类似的举动，这个 Issue 可能会被直接关闭

### 官方版原 Issue 编号

- Meloong-Git/PCL2#8018

### 描述

如同引用的上游 issue ，只有在令牌过期时才获取新的令牌

### 原因

每次打开启动器都要获取新的令牌，会拖延启动时间，部分网络差的环境时会更长，虽然会提示获取错误可忽略进入游戏

### LINKED ISSUE #3011: 重构 IdentityModel
此 issue 用于标记任务

解决 IdentityModel 的历史遗留问题，例如过于依赖 StandardExceptionGroup 导致异常语义不够明确

### LINKED ISSUE #3559: [C#]: 正版验证可被绕过
### 检查项 / Checklist

- [x] 我已在 [Issues 页面](https://github.com/PCL-Community/PCL-CE/issues?q=is%3Aissue+) 中搜索，确认了这一 Bug 未被提交过 / I have searched the [Issues page](https://github.com/PCL-Community/PCL-CE/issues?q=is%3Aissue+) and confirmed that this bug has not been reported yet.
- [ ] 我确认只是全部选中而没有[仔细确认](https://github.com/PCL-Community/PCL-CE/issues?q=is%3Aissue+)就直接提交了这个 Issue，并且同意这个 Issue 可以直接被关闭 / I confirm that I checked all boxes without carefully reading them, and agree that this issue may be closed directly.
- [x] 我确认正在使用当前更新分支的最新版 PCL CE，若不是，此 Issue 可能会被直接关闭（更新方式：启动器 `设置-关于-软件更新` 检查更新，或在 [Releases](https://github.com/PCL-Community/PCL-CE/releases) 页面下载） / I confirm that I am using the latest PCL CE version on the current update channel. If not, this issue may be closed directly. To update, check for updates in `Settings → About → Updates`, or download the latest version from the [Releases page](https://github.com/PCL-Community/PCL-CE/releases).
- [x] 我确认会详细描述我所遇到的问题，并附上日志和截图，而不是随便填点东西敷衍了事 / I will describe the problem in detail, attach logs and screenshots, and avoid submitting vague or incomplete information.

### 描述 / Description

如果先下载一个实例，并且去实例设置里面使用第三方登录就可以成功添加档案，并绕过正版验证

### 重现步骤 / Steps to Reproduce

1，下载一个实例
2，在实例设置里面，使用第三方如littleskin登录
3，登录成功后就添加档案了，可以启动游戏，而不需要正版验证

### 日志与附件 / Logs & Attachments

Image

Image

### PR DESCRIPTION
**Pull request by LuLu-ling**
> Partially generated by GPT-5.6-Sol

Close #2867 #17 #3011 #3559
## Sourcery 摘要

使用统一、持久化的个人资料服务和基于提供程序的身份验证架构，替换旧版 Minecraft 个人资料系统。

新功能：
- 添加统一的个人资料管理和身份验证支持，涵盖 Microsoft、Authlib、离线和 Yggdrasil Connect 账户。
- 添加 Yggdrasil Connect OAuth 设备登录、个人资料选择、服务器发现和身份验证服务器选择界面。
- 将个人资料持久化保存到加密配置中，并支持旧版个人资料迁移和稳定的个人资料标识。

错误修复：
- 通过强制使用 HTTPS 端点、验证签发者元数据和签名算法，以及拒绝未来签发的令牌，强化 OpenID 和 JWT 的处理。
- 修正 Yggdrasil 请求语义、验证端点、个人资料纹理 JSON 映射以及 PKCE 验证器处理。
- 通过稳定的标识跟踪，避免个人资料选择、更新、删除和持久化过程中的不一致问题。

增强功能：
- 将启动器个人资料状态、身份验证流程和凭据刷新逻辑从旧版 UI 模块迁移到可复用的 Core 个人资料服务和提供程序中。
- 改进 OAuth 请求构造、设备代码轮询、元数据默认值和响应模型覆盖范围。
- 在启动、皮肤管理、个人资料编辑和账户 UI 流程中全面集成新的个人资料服务。

测试：
- 添加个人资料管理测试，覆盖序列化往返、稳定的选择和更新、删除行为以及重复标识拒绝。

杂项：
- 在将旧版功能迁移到 Core 个人资料服务和 UI 适配器后，移除旧版 ModProfile 实现。

<details>
<summary>Original summary in English</summary>

## Summary by Sourcery

Replace the legacy Minecraft profile system with a unified, persistent profile service and provider-based authentication architecture.

New Features:
- Add unified profile management and authentication support for Microsoft, Authlib, offline, and Yggdrasil Connect accounts.
- Add Yggdrasil Connect OAuth device login, profile selection, server discovery, and authentication-server selection UI.
- Persist profiles in encrypted configuration with legacy profile migration and stable profile identities.

Bug Fixes:
- Harden OpenID and JWT handling by enforcing HTTPS endpoints, validating issuer metadata and signing algorithms, and rejecting tokens issued in the future.
- Correct Yggdrasil request semantics, validation endpoints, profile texture JSON mappings, and PKCE verifier handling.
- Prevent profile selection, update, deletion, and persistence inconsistencies through stable identity tracking.

Enhancements:
- Move launcher profile state, authentication workflows, and credential refresh logic from the legacy UI module into reusable Core profile services and providers.
- Improve OAuth request construction, device-code polling, metadata defaults, and response-model coverage.
- Integrate the new profile service throughout launching, skin management, profile editing, and account UI flows.

Tests:
- Add profile management tests covering serialization round trips, stable selection and updates, deletion behavior, and duplicate identity rejection.

Chores:
- Remove the legacy ModProfile implementation after migrating its functionality to the Core profile service and UI adapter.

</details>

### CODE REVIEW

File: PCL.Core/Minecraft/Profile/Authentication/YggdrasilConnectProvider.cs [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+        {
+            throw;
+        }
+        catch
+        {
+            return null;
+        }
+    }
+
+    private void EnsureInitialized()
```
Comment (whitecat346) [inline PCL.Core/Minecraft/Profile/Authentication/YggdrasilConnectProvider.cs]: 不符合命名规范，改为`_EnsureInitialized`

File: PCL.Core/Minecraft/Profile/Authentication/YggdrasilConnectProvider.cs [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+    public static bool TryGetBuiltInClientId(string host, out string clientId)
+    {
+        var normalizedHost = host.Trim().TrimEnd('.');
+        return _BuiltInClientIds.TryGetValue(normalizedHost, out clientId!);
+    }
+
+    public async Task InitializeAsync(CancellationToken token)
+    {
+        _client = new YggdrasilClient(_options);
+        await _client.InitializeAsync(token).ConfigureAwait(false);
```
Comment (whitecat346) [inline PCL.Core/Minecraft/Profile/Authentication/YggdrasilConnectProvider.cs]:
```suggestion
        return _client.InitializeAsync(token);
```
这个我认为是个转发，用await没有什么意义

File: Plain Craft Launcher 2/Modules/ModLink.cs [hunk 3]
```diff
@@ -28,8 +29,8 @@ public static bool LobbyPrecheck()
             return false;
         }
 
-        if (ModProfile.selectedProfile is not null)
-            if (ModProfile.selectedProfile.Username.Contains("|"))
+        if (ProfileService.Current is not null)
+            if (ProfileService.Current.UserName.Contains("|"))
```
Comment (whitecat346) [inline Plain Craft Launcher 2/Modules/ModLink.cs]:
```suggestion
            if (ProfileService.Current.UserName.Contains('|'))
```
避免Culture问题

File: Plain Craft Launcher 2/Modules/ModMain.cs [hunk 4]
```diff
@@ -544,6 +557,33 @@ public static string MyMsgBoxInput(string title, string text = "", string defaul
...
+        {
+            frmMain?.DragStop();
+            ComponentDispatcher.PushModal();
+            Dispatcher.PushFrame(converter.WaitFrame);
+        }
+        finally
+        {
+            ComponentDispatcher.PopModal();
+        }
+        return converter.Result?.ToString();
```
Comment (whitecat346) [inline Plain Craft Launcher 2/Modules/ModMain.cs]: 可能返回null，加个回退值

File: Plain Craft Launcher 2/FormMain.xaml.cs [hunk 5]
```diff
@@ -519,7 +520,7 @@ public void EndProgram(bool sendWarning, bool isUpdating = false)
         // 关闭联机大厅
         // Await LobbyController.CloseAsync().ConfigureAwait(False)
         // 存储上次使用的档案编号
-        ModProfile.SaveProfile();
+        ProfileService.Save();
```
Comment (whitecat346) [inline Plain Craft Launcher 2/FormMain.xaml.cs]: 在`ProfileService`的`_Stop()`逻辑里已经有了`Save()`的操作，这里是多此一举

**Review by whitecat346 — CHANGES_REQUESTED** [review]
- 对于PCL下的方法，很多遇到了异常都是直接抛`Exception`，这不是一个好的实践，不利于上层调用通过try-catch确定问题类型
- 其余详见comments

File: Plain Craft Launcher 2/Modules/Minecraft/ProfileUi.cs [hunk 6]
```diff
@@ -0,0 +1,520 @@
...
+
+/// <summary>
+/// WPF-only profile workflows. Profile state and persistence live in the Core ProfileService.
+/// </summary>
+public static class ProfileUi
+{
+    private static int _isMsSkinChanging;
+
+    public static void ProfileLog(string content, ModBase.LogLevel level = ModBase.LogLevel.Normal)
+        => ModBase.Log("[Profile] " + content, level);
```
Comment (Hill23333) [inline Plain Craft Launcher 2/Modules/Minecraft/ProfileUi.cs]: 应当使用 LogWrapper 提供的方法而非自建 Wrapper

**Review by Hill23333 — DISMISSED** [review]

**Review by whitecat346 — DISMISSED** [review]

**Review by MoYuan-CN — DISMISSED** [review]

**Review by whitecat346 — APPROVED** [review]
LGTM

**Review by MoYuan-CN — APPROVED** [review]
